wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheMonitor.java

```diff
@@ -0,0 +1,917 @@
+          long targetInterval = hasActiveMonitoring
+              ? TimeUnit.SECONDS.toMillis(CACHE_HEALTH_CHECK_INTERVAL)
+              : THREAD_SLEEP_WHEN_INACTIVE_MILLIS;
+          long sleepMillis = Math.max(0, targetInterval - iterationDuration);
+          sleep(sleepMillis);
+        } catch (final InterruptedException intEx) {
+          throw intEx;
+        } catch (final Exception ex) {
+          LOGGER.warning("Exception during cache monitoring: " + ex.getMessage());
+        }
+      }
+    } catch (final InterruptedException intEx) {
+      LOGGER.warning("Cache monitor thread interrupted");
+    } finally {
+      this.stopped = true;
+      LOGGER.info("Cache monitor thread stopped");
+    }
+  }
+
+  /**
+   * Executes a ping operation to check cache endpoint health.
+   */
+  private boolean executePing(ClusterHealthState cluster, boolean isRw) {
+    String endpoint = isRw ? cluster.rwEndpoint : cluster.roEndpoint;
+    long lastPingTimestamp = isRw ? cluster.rwLastPingTimestamp : cluster.roLastPingTimestamp;
+
+    try {
+      long currentTime = System.currentTimeMillis();
+      long timeSinceLastPing = lastPingTimestamp == 0 ? 0 : currentTime - lastPingTimestamp;
+
+      String humanReadableTime = formatTimestamp(currentTime);
+
+      // Log ping start
+      if (lastPingTimestamp == 0) {
+        LOGGER.info(String.format(
+            "[PING_START] Endpoint: %s (%s) | Time: %s | First ping for this endpoint",
+            endpoint, isRw ? "RW" : "RO", humanReadableTime));
+      } else {
+        LOGGER.info(String.format(
+            "[PING_START] Endpoint: %s (%s) | Time: %s | Interval since last: %dms (%.2fs)",
+            endpoint, isRw ? "RW" : "RO", humanReadableTime, timeSinceLastPing, timeSinceLastPing / 1000.0));
+      }
+
+      long pingStartTime = System.currentTimeMillis();
+      boolean result = ping(cluster, isRw);
+      long pingDuration = System.currentTimeMillis() - pingStartTime;
+
+      HealthState currentState = isRw ? cluster.rwHealthState : cluster.roHealthState;
+      LOGGER.info(String.format(
+          "[PING_RESULT] Endpoint: %s (%s) | Result: %s | Ping duration: %dms | State: %s",
+          endpoint, isRw ? "RW" : "RO", result ? "SUCCESS" : "FAILURE", pingDuration, currentState));
+
+      // Update last ping timestamp
+      if (isRw) {
+        cluster.rwLastPingTimestamp = currentTime;
+      } else {
+        cluster.roLastPingTimestamp = currentTime;
+      }
+
+      return result;
+    } catch (Exception e) {
+      LOGGER.warning(String.format(
+          "[PING_ERROR] Endpoint: %s (%s) | Error: %s",
+          endpoint, isRw ? "RW" : "RO", e.getMessage()));
+      return false;
+    }
+  }
+
+  /**
+   * Handles the result of a ping operation and updates cluster state accordingly.
+   */
+  private void handlePingResult(ClusterHealthState cluster, boolean isRw, boolean pingSuccess) {
+    String endpoint = isRw ? cluster.rwEndpoint : cluster.roEndpoint;
```

Comment:
You don't need to determine which endpoint you are pinging if this code is in the same method as `executePing()`.